core/src/main/java/nl/nn/adapterframework/jwt/JwtValidator.java

```diff
@@ -0,0 +1,145 @@
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package nl.nn.adapterframework.jwt;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.text.ParseException;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jose.jwk.source.ImmutableJWKSet;
+import com.nimbusds.jose.jwk.source.JWKSource;
+import com.nimbusds.jose.jwk.source.RemoteJWKSet;
+import com.nimbusds.jose.proc.BadJOSEException;
+import com.nimbusds.jose.proc.JWSKeySelector;
+import com.nimbusds.jose.proc.JWSVerificationKeySelector;
+import com.nimbusds.jose.proc.SecurityContext;
+import com.nimbusds.jose.util.DefaultResourceRetriever;
+import com.nimbusds.jose.util.ResourceRetriever;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.proc.BadJWTException;
+import com.nimbusds.jwt.proc.ConfigurableJWTProcessor;
+import com.nimbusds.jwt.proc.DefaultJWTClaimsVerifier;
+import com.nimbusds.jwt.proc.DefaultJWTProcessor;
+
+import lombok.Getter;
+
+public class JwtValidator<C extends SecurityContext> {
+
+	private @Getter int connectTimeout=2000;
+	private @Getter int readTimeout=2000;
+
+	private @Getter ConfigurableJWTProcessor<C> jwtProcessor = null;
+
+	public JwtValidator() {
+		jwtProcessor = new DefaultJWTProcessor<C>();
+	}
+
+	public void init(String jwksUrl, String requiredIssuer, String requiredClaims, String exactMatchClaims) throws ParseException, MalformedURLException, IOException {
+		JWKSource<C> keySource = getKeySource(new URL(jwksUrl));
+
+		// The expected JWS algorithm of the access tokens (agreed out-of-band)
+		JWSAlgorithm expectedJWSAlg = JWSAlgorithm.RS256;
+
+		// Configure the JWT processor with a key selector to feed matching public
+		// RSA keys sourced from the JWK set URL
+		JWSKeySelector<C> keySelector = new JWSVerificationKeySelector<C>(expectedJWSAlg, keySource);
+
+		Set<String> requiredClaimsSet = null;
+		if(StringUtils.isNotEmpty(requiredClaims)) {
+			requiredClaimsSet = Stream.of(requiredClaims.split("\\s*,\\s*"))
+									.map(String::trim)
+									.collect(Collectors.toSet());
+		}
+
+		JWTClaimsSet exactMatchClaimsSet = null;
+		if(StringUtils.isNotEmpty(exactMatchClaims)) {
+			Map<String, Object> claimsMap = Stream.of(exactMatchClaims.split("\\s*,\\s*"))
+												.map(s -> s.split("\\s*=\\s*"))
+												.collect(Collectors.toMap(a -> a[0], a -> a.length > 1 ? a[1] : ""));
+			exactMatchClaimsSet = JWTClaimsSet.parse(claimsMap);
+		}
```

Comment:
Where do you do the verification of requiredClaims and exactMatchClaims? The local variables are not used.
I would expect that to be done somewhere else:
Failure to validate the token would result in a HTTP status 401
Failure to comply to the claims would result in a HTTP status 403

Reply:
requiredClaims and exactMatchClaims are used to initialize the verifier. The verifier is a part of jwtProcessor. By calling process method of the processor with the token claims will be verified. In case verification fails 401 is returned